File: backend/app/services/onboarding_service.py
import psycopg2
from psycopg2.extras import RealDictCursor
import logging
from flask import current_app
from urllib.parse import urlparse
from ..services.database_service import DatabaseService

logger = logging.getLogger(__name__)

class OnboardingService:
    @staticmethod
    def save_step1_data(firebase_uid, step1_data):
        """Save or update Step 1 onboarding data."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            # Prepare the data
            full_name = step1_data.get('fullName')
            dob = step1_data.get('dob') if step1_data.get('dob') else None
            address = step1_data.get('address')
            email = step1_data.get('email')
            phone_number = step1_data.get('phoneNumber')
            nz_residency_status = step1_data.get('nzResidencyStatus')
            tax_number = step1_data.get('taxNumber')
            
            # Use UPSERT (INSERT ... ON CONFLICT) to handle both insert and update
            cursor.execute("""
                INSERT INTO onboarding_applications (
                    firebase_uid, full_name, date_of_birth, address, email, 
                    phone_number, nz_residency_status, tax_number, step_completed
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (firebase_uid) 
                DO UPDATE SET
                    full_name = EXCLUDED.full_name,
                    date_of_birth = EXCLUDED.date_of_birth,
                    address = EXCLUDED.address,
                    email = EXCLUDED.email,
                    phone_number = EXCLUDED.phone_number,
                    nz_residency_status = EXCLUDED.nz_residency_status,
                    tax_number = EXCLUDED.tax_number,
                    step_completed = GREATEST(onboarding_applications.step_completed, 1),
                    updated_at = CURRENT_TIMESTAMP
                RETURNING id, firebase_uid, full_name, email, step_completed, created_at, updated_at;
            """, (
                firebase_uid, full_name, dob, address, email, 
                phone_number, nz_residency_status, tax_number, 1
            ))
            
            result = cursor.fetchone()
            conn.commit()
            cursor.close()
            conn.close()
            
            return True, dict(result)
            
        except Exception as e:
            logger.error(f"Failed to save Step 1 data: {e}")
            return False, str(e)

    @staticmethod
    def save_step2_data(firebase_uid, step2_data):
        """Save or update Step 2 onboarding data."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            # Prepare the data
            employment_type = step2_data.get('employmentType')
            employer = step2_data.get('employer')
            job_title = step2_data.get('jobTitle')
            employment_duration = step2_data.get('employmentDuration')
            monthly_income = step2_data.get('monthlyIncome')
            other_income = step2_data.get('otherIncome', 0)
            
            # Use UPSERT to handle both insert and update
            cursor.execute("""
                INSERT INTO onboarding_applications (
                    firebase_uid, employment_type, employer, job_title, 
                    employment_duration, monthly_income, other_income, step_completed
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (firebase_uid) 
                DO UPDATE SET
                    employment_type = EXCLUDED.employment_type,
                    employer = EXCLUDED.employer,
                    job_title = EXCLUDED.job_title,
                    employment_duration = EXCLUDED.employment_duration,
                    monthly_income = EXCLUDED.monthly_income,
                    other_income = EXCLUDED.other_income,
                    step_completed = GREATEST(onboarding_applications.step_completed, 2),
                    updated_at = CURRENT_TIMESTAMP
                RETURNING id, firebase_uid, employment_type, employer, job_title, 
                         employment_duration, monthly_income, other_income, 
                         step_completed, created_at, updated_at;
            """, (
                firebase_uid, employment_type, employer, job_title, 
                employment_duration, monthly_income, other_income, 2
            ))
            
            result = cursor.fetchone()
            conn.commit()
            cursor.close()
            conn.close()
            
            return True, dict(result)
            
        except Exception as e:
            logger.error(f"Failed to save Step 2 data: {e}")
            return False, str(e)

    @staticmethod
    def get_step1_data(firebase_uid):
        """Get Step 1 onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT 
                    id, firebase_uid, full_name, date_of_birth, address, email,
                    phone_number, nz_residency_status, tax_number, step_completed,
                    is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s;
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                # Convert to frontend-compatible format
                data = dict(result)
                # Convert date to string format for frontend
                if data['date_of_birth']:
                    data['date_of_birth'] = data['date_of_birth'].strftime('%Y-%m-%d')
                
                return True, data
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get Step 1 data: {e}")
            return False, str(e)

    @staticmethod
    def get_step2_data(firebase_uid):
        """Get Step 2 onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT 
                    id, firebase_uid, employment_type, employer, job_title,
                    employment_duration, monthly_income, other_income, 
                    step_completed, is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s;
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                return True, dict(result)
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get Step 2 data: {e}")
            return False, str(e)

    @staticmethod
    def get_user_onboarding_status(firebase_uid):
        """Get user's onboarding completion status."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT step_completed, is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s;
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                return True, dict(result)
            else:
                return True, {'step_completed': 0, 'is_completed': False}
                
        except Exception as e:
            logger.error(f"Failed to get onboarding status: {e}")
            return False, str(e)

    @staticmethod
    def get_complete_user_data(firebase_uid):
        """Get all onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT *
                FROM onboarding_applications 
                WHERE firebase_uid = %s;
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                data = dict(result)
                # Convert date to string format for frontend
                if data.get('date_of_birth'):
                    data['date_of_birth'] = data['date_of_birth'].strftime('%Y-%m-%d')
                
                return True, data
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get complete user data: {e}")
            return False, str(e)
        

    @staticmethod
    def save_step3_data(firebase_uid, step3_data):
        """Save or update Step 3 onboarding data."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            # Prepare the data
            rent = step3_data.get('rent')
            monthly_expenses = step3_data.get('monthlyExpenses')
            debts = step3_data.get('debts')
            dependents = step3_data.get('dependents')
            
            # Validate that at least one field is provided
            if all(v is None for v in [rent, monthly_expenses, debts, dependents]):
                return False, "At least one Step 3 field must be provided"
            
            # Use UPSERT to handle both insert and update
            cursor.execute("""
                INSERT INTO onboarding_applications (
                    firebase_uid, rent, monthly_expenses, debts, dependents, step_completed
                ) VALUES (%s, %s, %s, %s, %s, %s)
                ON CONFLICT (firebase_uid) 
                DO UPDATE SET
                    rent = EXCLUDED.rent,
                    monthly_expenses = EXCLUDED.monthly_expenses,
                    debts = EXCLUDED.debts,
                    dependents = EXCLUDED.dependents,
                    step_completed = GREATEST(onboarding_applications.step_completed, 3),
                    updated_at = CURRENT_TIMESTAMP
                RETURNING id, firebase_uid, rent, monthly_expenses, debts, dependents,
                        step_completed, created_at, updated_at;
            """, (
                firebase_uid, rent, monthly_expenses, debts, dependents, 3
            ))
            
            result = cursor.fetchone()
            conn.commit()
            cursor.close()
            conn.close()
            
            return True, dict(result)
            
        except Exception as e:
            logger.error(f"Failed to save Step 3 data: {e}")
            return False, str(e)

    @staticmethod
    def get_step3_data(firebase_uid):
        """Get Step 3 onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT 
                    id, firebase_uid, rent, monthly_expenses, debts, dependents,
                    step_completed, is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                return True, dict(result)
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get Step 3 data: {e}")
            return False, str(e)

    

    @staticmethod
    def save_step4_data(firebase_uid, step4_data):
        """Save or update Step 4 onboarding data."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            # Prepare the data
            savings = step4_data.get('savings')
            assets = step4_data.get('assets')
            source_of_funds = step4_data.get('sourceOfFunds')
            expected_account_activity = step4_data.get('expectedAccountActivity')
            is_politically_exposed = step4_data.get('isPoliticallyExposed')
            
            # Use UPSERT to handle both insert and update
            cursor.execute("""
                INSERT INTO onboarding_applications (
                    firebase_uid, savings, assets, source_of_funds, 
                    expected_account_activity, is_politically_exposed, step_completed
                ) VALUES (%s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (firebase_uid) 
                DO UPDATE SET
                    savings = EXCLUDED.savings,
                    assets = EXCLUDED.assets,
                    source_of_funds = EXCLUDED.source_of_funds,
                    expected_account_activity = EXCLUDED.expected_account_activity,
                    is_politically_exposed = EXCLUDED.is_politically_exposed,
                    step_completed = GREATEST(onboarding_applications.step_completed, 4),
                    updated_at = CURRENT_TIMESTAMP
                RETURNING id, firebase_uid, savings, assets, source_of_funds,
                        expected_account_activity, is_politically_exposed,
                        step_completed, created_at, updated_at;
            """, (
                firebase_uid, savings, assets, source_of_funds, 
                expected_account_activity, is_politically_exposed, 4
            ))
            
            result = cursor.fetchone()
            conn.commit()
            cursor.close()
            conn.close()
            
            return True, dict(result)
            
        except Exception as e:
            logger.error(f"Failed to save Step 4 data: {e}")
            return False, str(e)

    @staticmethod
    def get_step4_data(firebase_uid):
        """Get Step 4 onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT 
                    id, firebase_uid, savings, assets, source_of_funds,
                    expected_account_activity, is_politically_exposed,
                    step_completed, is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                return True, dict(result)
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get Step 4 data: {e}")
            return False, str(e)

    @staticmethod
    def save_step5_data(firebase_uid, step5_data):
        """Save or update Step 5 onboarding data."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            # Prepare the data
            loan_amount = step5_data.get('loanAmount')
            loan_purpose = step5_data.get('loanPurpose')
            loan_term = step5_data.get('loanTerm')
            understands_terms = step5_data.get('understandsTerms', False)
            can_afford_repayments = step5_data.get('canAffordRepayments', False)
            has_received_advice = step5_data.get('hasReceivedAdvice', False)
            
            # Use UPSERT to handle both insert and update
            cursor.execute("""
                INSERT INTO onboarding_applications (
                    firebase_uid, loan_amount, loan_purpose, loan_term,
                    understands_terms, can_afford_repayments, has_received_advice, step_completed
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (firebase_uid) 
                DO UPDATE SET
                    loan_amount = EXCLUDED.loan_amount,
                    loan_purpose = EXCLUDED.loan_purpose,
                    loan_term = EXCLUDED.loan_term,
                    understands_terms = EXCLUDED.understands_terms,
                    can_afford_repayments = EXCLUDED.can_afford_repayments,
                    has_received_advice = EXCLUDED.has_received_advice,
                    step_completed = GREATEST(onboarding_applications.step_completed, 5),
                    updated_at = CURRENT_TIMESTAMP
                RETURNING id, firebase_uid, loan_amount, loan_purpose, loan_term,
                        understands_terms, can_afford_repayments, has_received_advice,
                        step_completed, created_at, updated_at;
            """, (
                firebase_uid, loan_amount, loan_purpose, loan_term,
                understands_terms, can_afford_repayments, has_received_advice, 5
            ))
            
            result = cursor.fetchone()
            conn.commit()
            cursor.close()
            conn.close()
            
            return True, dict(result)
            
        except Exception as e:
            logger.error(f"Failed to save Step 5 data: {e}")
            return False, str(e)

    @staticmethod
    def get_step5_data(firebase_uid):
        """Get Step 5 onboarding data for a user."""
        try:
            conn = DatabaseService.get_connection()
            cursor = conn.cursor()
            
            cursor.execute("""
                SELECT 
                    id, firebase_uid, loan_amount, loan_purpose, loan_term,
                    understands_terms, can_afford_repayments, has_received_advice,
                    step_completed, is_completed, created_at, updated_at
                FROM onboarding_applications 
                WHERE firebase_uid = %s
            """, (firebase_uid,))
            
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            
            if result:
                return True, dict(result)
            else:
                return True, None
                
        except Exception as e:
            logger.error(f"Failed to get Step 5 data: {e}")
            return False, str(e)

File: frontend/lib/services/onboardingService.ts
import { apiClient } from '@/lib/firebase/api';
import { 
  Step1Data,
  Step1ApiData,
  Step1ResponseData,
  Step2Data,
  Step2ApiData,
  Step2ResponseData,
  Step3Data,
  Step3ApiData,
  Step4Data,
  Step4ApiData,
  Step5ApiData,
  Step5Data,
  Step6Data,
  Step6ApiData,
  Step6ResponseData,
  Step5ResponseData,
  Step4ResponseData,
  Step3ResponseData,
  OnboardingStatusData,
  OnboardingStatus,
  SavedStep1Data,
  SavedStep2Data,
  SavedStep3Data,
  SavedStep4Data,
  SavedStep5Data,
  SavedStep6Data
} from '@/types/onboarding';







export class OnboardingService {
  /**
   * Initialize onboarding tables (call this once when app starts)
   */
  static async initializeOnboarding(): Promise<void> {
    try {
      await apiClient.post('/api/onboarding/initialize');
    } catch (error) {
      console.error('Failed to initialize onboarding:', error);
      throw error;
    }
  }

  /**
   * Convert domain model to API format
   */
  private static toApiFormat(data: Step1Data): Step1ApiData {
    return {
      fullName: data.fullName,
      dob: data.dob,
      address: data.address,
      email: data.email,
      phoneNumber: data.phoneNumber,
      nzResidencyStatus: data.nzResidencyStatus,
      ...(data.taxNumber && { taxNumber: data.taxNumber }),
    };
  }

  /**
   * Convert API response to domain model
   */
  private static fromApiFormat(data: Step1ResponseData): SavedStep1Data {
    return {
      fullName: data.fullName,
      dob: data.dob,
      address: data.address,
      email: data.email,
      phoneNumber: data.phoneNumber,
      nzResidencyStatus: data.nzResidencyStatus,
      taxNumber: data.taxNumber,
      stepCompleted: data.stepCompleted,
      isCompleted: data.isCompleted,
    };
  }

  /**
   * Save Step 1 data to the backend
   */
  static async saveStep1Data(data: Step1Data): Promise<SavedStep1Data> {
    try {
      const apiData = this.toApiFormat(data);
      
      const response = await apiClient.post<Step1ResponseData, Step1ApiData>(
        '/api/onboarding/step1', 
        apiData
      );
      
      return this.fromApiFormat(response);
    } catch (error) {
      console.error('Failed to save Step 1 data:', error);
      throw error;
    }
  }
  
  // Step 2 Methods
  /**
   * Convert Step 2 domain model to API format
   */
  private static toStep2ApiFormat(data: Step2Data): Step2ApiData {
    return {
      employmentType: data.employmentType,
      ...(data.employer && { employer: data.employer }),
      ...(data.jobTitle && { jobTitle: data.jobTitle }),
      ...(data.employmentDuration && { employmentDuration: data.employmentDuration }),
      ...(data.monthlyIncome !== undefined && { monthlyIncome: data.monthlyIncome }),
      ...(data.otherIncome !== undefined && { otherIncome: data.otherIncome }),
    };
  }
   /**
   * Convert Step 2 API response to domain model
   */
   private static fromStep2ApiFormat(data: Step2ResponseData): SavedStep2Data {
    return {
      employmentType: data.employmentType as Step2Data['employmentType'],
      employer: data.employer,
      jobTitle: data.jobTitle,
      employmentDuration: data.employmentDuration,
      monthlyIncome: data.monthlyIncome,
      otherIncome: data.otherIncome,
      stepCompleted: data.stepCompleted,
      isCompleted: data.isCompleted,
    };
  }

  static async saveStep2Data(data: Step2Data): Promise<SavedStep2Data> {
    try {
      const apiData = this.toStep2ApiFormat(data);
      
      const response = await apiClient.post<Step2ResponseData, Step2ApiData>(
        '/api/onboarding/step2', 
        apiData
      );
      
      return this.fromStep2ApiFormat(response);
    } catch (error) {
      console.error('Failed to save Step 2 data:', error);
      throw error;
    }
  }

  /**
   * Get Step 1 data from the backend
   */
  static async getStep1Data(): Promise<SavedStep1Data | null> {
    try {
      const response = await apiClient.get<Step1ResponseData>('/api/onboarding/step1');
      
      if (!response) {
        return null;
      }
      
      return this.fromApiFormat(response);
    } catch (error) {
      console.error('Failed to get Step 1 data:', error);
      // Return null if no data found (user hasn't started onboarding)
      if (error instanceof Error && error.message.includes('No Step 1 data found')) {
        return null;
      }
      throw error;
    }
  }

  static async getStep2Data(): Promise<SavedStep2Data | null> {
    try {
      const response = await apiClient.get<Step2ResponseData>('/api/onboarding/step2');
      
      if (!response) {
        return null;
      }
      
      return this.fromStep2ApiFormat(response);
    } catch (error) {
      console.error('Failed to get Step 2 data:', error);
      // Return null if no data found (user hasn't started step 2)
      if (error instanceof Error && error.message.includes('No Step 2 data found')) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Get user's onboarding status
   */
  static async getOnboardingStatus(): Promise<OnboardingStatus> {
    try {
      const response = await apiClient.get<OnboardingStatusData>('/api/onboarding/status');
      
      return {
        stepCompleted: response.stepCompleted,
        isCompleted: response.isCompleted,
        created_at: response.created_at,
        updated_at: response.updated_at,
      };
    } catch (error) {
      console.error('Failed to get onboarding status:', error);
      throw error;
    }
  }

  // Add these methods to your frontend/lib/services/onboardingService.ts file

/**
 * Convert Step 3 domain model to API format
 */
private static toStep3ApiFormat(data: Step3Data): Step3ApiData {
  return {
    rent: data.rent,
    monthlyExpenses: data.monthlyExpenses,
    debts: data.debts,
    dependents: data.dependents,
  };
}

  /**
   * Convert Step 3 API response to domain model
   */
  private static fromStep3ApiFormat(data: Step3ResponseData): SavedStep3Data {
    return {
      rent: data.rent,
      monthlyExpenses: data.monthlyExpenses,
      debts: data.debts,
      dependents: data.dependents,
      stepCompleted: data.stepCompleted,
      isCompleted: data.isCompleted,
    };
  }

  /**
   * Save Step 3 data to the backend
   */
  static async saveStep3Data(data: Step3Data): Promise<SavedStep3Data> {
    try {
      const apiData = this.toStep3ApiFormat(data);
      
      const response = await apiClient.post<Step3ResponseData, Step3ApiData>(
        '/api/onboarding/step3',
        apiData
      );
      
      if (!response) {
        throw new Error('No response received from server');
      }
      
      return this.fromStep3ApiFormat(response);
    } catch (error) {
      console.error('Failed to save Step 3 data:', error);
      throw error;
    }
  }

  /**
   * Get Step 3 data from the backend
   */
  static async getStep3Data(): Promise<SavedStep3Data | null> {
    try {
      const response = await apiClient.get<Step3ResponseData>('/api/onboarding/step3');
      
      if (!response) {
        return null;
      }
      
      return this.fromStep3ApiFormat(response);
    } catch (error) {
      console.error('Failed to get Step 3 data:', error);
      // Return null if no data found (user hasn't started step 3)
      if (error instanceof Error && error.message.includes('No Step 3 data found')) {
        return null;
      }
      throw error;
    }
  }

 // Step 4 Methods
// Add Step4Data, Step4ApiData, Step4ResponseData, SavedStep4Data to your imports

// Step 4 Methods - copy Step 2 pattern exactly
/**
 * Convert Step 4 domain model to API format
 */
private static toStep4ApiFormat(data: Step4Data): Step4ApiData {
  return {
    savings: data.savings,
    assets: data.assets,
    sourceOfFunds: data.sourceOfFunds,
    expectedAccountActivity: data.expectedAccountActivity,
    isPoliticallyExposed: data.isPoliticallyExposed,
  };
}

/**
 * Convert Step 4 API response to domain model
 */
private static fromStep4ApiFormat(data: Step4ResponseData): SavedStep4Data {
  return {
    savings: data.savings,
    assets: data.assets,
    sourceOfFunds: data.sourceOfFunds,
    expectedAccountActivity: data.expectedAccountActivity,
    isPoliticallyExposed: data.isPoliticallyExposed,
    stepCompleted: data.stepCompleted,
    isCompleted: data.isCompleted,
  };
}

static async saveStep4Data(data: Step4Data): Promise<SavedStep4Data> {
  try {
    const apiData = this.toStep4ApiFormat(data);
    
    const response = await apiClient.post<Step4ResponseData, Step4ApiData>(
      '/api/onboarding/step4', 
      apiData
    );
    
    return this.fromStep4ApiFormat(response);
  } catch (error) {
    console.error('Failed to save Step 4 data:', error);
    throw error;
  }
}

static async getStep4Data(): Promise<SavedStep4Data | null> {
  try {
    const response = await apiClient.get<Step4ResponseData>('/api/onboarding/step4');
    
    if (!response) {
      return null;
    }
    
    return this.fromStep4ApiFormat(response);
  } catch (error) {
    console.error('Failed to get Step 4 data:', error);
    if (error instanceof Error && error.message.includes('No Step 4 data found')) {
      return null;
    }
    throw error;
  }
}

/**
 * Convert Step 5 domain model to API format
 */
private static toStep5ApiFormat(data: Step5Data): Step5ApiData {
  return {
    loanAmount: data.loanAmount,
    loanPurpose: data.loanPurpose,
    loanTerm: data.loanTerm,
    understandsTerms: data.understandsTerms,
    canAffordRepayments: data.canAffordRepayments,
    hasReceivedAdvice: data.hasReceivedAdvice,
  };
}

/**
 * Convert Step 5 API response to domain model
 */
private static fromStep5ApiFormat(data: Step5ResponseData): SavedStep5Data {
  return {
    loanAmount: data.loanAmount,
    loanPurpose: data.loanPurpose || '',
    loanTerm: data.loanTerm,
    understandsTerms: data.understandsTerms,
    canAffordRepayments: data.canAffordRepayments,
    hasReceivedAdvice: data.hasReceivedAdvice,
    stepCompleted: data.stepCompleted,
    isCompleted: data.isCompleted,
  };
}

/**
 * Save Step 5 data to the backend
 */
static async saveStep5Data(data: Step5Data): Promise<SavedStep5Data> {
  try {
    const apiData = this.toStep5ApiFormat(data);
    
    const response = await apiClient.post<Step5ResponseData, Step5ApiData>(
      '/api/onboarding/step5',
      apiData
    );
    
    if (!response) {
      throw new Error('No response received from server');
    }
    
    return this.fromStep5ApiFormat(response);
  } catch (error) {
    console.error('Failed to save Step 5 data:', error);
    throw error;
  }
}

/**
 * Get Step 5 data from the backend
 */
static async getStep5Data(): Promise<SavedStep5Data | null> {
  try {
    const response = await apiClient.get<Step5ResponseData>('/api/onboarding/step5');
    
    if (!response) {
      return null;
    }
    
    return this.fromStep5ApiFormat(response);
  } catch (error) {
    console.error('Failed to get Step 5 data:', error);
    // Return null if no data found (user hasn't started step 5)
    if (error instanceof Error && error.message.includes('No Step 5 data found')) {
      return null;
    }
    throw error;
  }
}

// Step 6 Methods - API Compatible
private static toStep6ApiFormat(data: Step6Data): Step6ApiData {
  return {
    identityDocumentName: data.identityDocumentName,
    identityDocumentSize: data.identityDocumentSize,
    identityDocumentType: data.identityDocumentType,
    identityDocumentUploadedAt: data.identityDocumentUploadedAt,
    addressProofName: data.addressProofName,
    addressProofSize: data.addressProofSize,
    addressProofType: data.addressProofType,
    addressProofUploadedAt: data.addressProofUploadedAt,
    incomeProofName: data.incomeProofName,
    incomeProofSize: data.incomeProofSize,
    incomeProofType: data.incomeProofType,
    incomeProofUploadedAt: data.incomeProofUploadedAt,
  };
}

private static fromStep6ApiFormat(data: Step6ResponseData): SavedStep6Data {
  return {
    identityDocumentName: data.identityDocumentName,
    identityDocumentSize: data.identityDocumentSize,
    identityDocumentType: data.identityDocumentType,
    identityDocumentUploadedAt: data.identityDocumentUploadedAt,
    addressProofName: data.addressProofName,
    addressProofSize: data.addressProofSize,
    addressProofType: data.addressProofType,
    addressProofUploadedAt: data.addressProofUploadedAt,
    incomeProofName: data.incomeProofName,
    incomeProofSize: data.incomeProofSize,
    incomeProofType: data.incomeProofType,
    incomeProofUploadedAt: data.incomeProofUploadedAt,
    stepCompleted: data.stepCompleted,
    isCompleted: data.isCompleted,
  };
}

static async saveStep6Data(data: Step6Data): Promise<SavedStep6Data> {
  try {
    const apiData = this.toStep6ApiFormat(data);
    
    const response = await apiClient.post<Step6ResponseData, Step6ApiData>(
      '/api/onboarding/step6', 
      apiData
    );
    
    return this.fromStep6ApiFormat(response);
  } catch (error) {
    console.error('Failed to save Step 6 data:', error);
    throw error;
  }
}

static async getStep6Data(): Promise<SavedStep6Data | null> {
  try {
    const response = await apiClient.get<Step6ResponseData>('/api/onboarding/step6');
    
    if (!response) {
      return null;
    }
    
    return this.fromStep6ApiFormat(response);
  } catch (error) {
    console.error('Failed to get Step 6 data:', error);
    if (error instanceof Error && error.message.includes('No Step 6 data found')) {
      return null;
    }
    throw error;
  }
}
 
}